// @triniwiz/nativescript-toasty/index.android.js
import { Application, Color, Screen, Utils } from '@nativescript/core';
import { ToastDuration, ToastPosition } from './common';
export { ToastDuration, ToastPosition } from './common';
export class Toasty {
    constructor(opts) {
        var _a, _b, _c, _d, _e, _f, _g;
        this._screenHeight = Screen.mainScreen.heightPixels;
        this._screenWidth = Screen.mainScreen.widthPixels;
        this._text = opts === null || opts === void 0 ? void 0 : opts.text;
        this._duration = (_a = opts === null || opts === void 0 ? void 0 : opts.duration) !== null && _a !== void 0 ? _a : ToastDuration.SHORT;
        this._position = (_b = opts === null || opts === void 0 ? void 0 : opts.position) !== null && _b !== void 0 ? _b : ToastPosition.BOTTOM;
        this._textColor = (_c = opts === null || opts === void 0 ? void 0 : opts.textColor) !== null && _c !== void 0 ? _c : 'white';
        this._backgroundColor = (_d = opts === null || opts === void 0 ? void 0 : opts.backgroundColor) !== null && _d !== void 0 ? _d : 'black';
        this._androidOpts = (_e = opts === null || opts === void 0 ? void 0 : opts.android) !== null && _e !== void 0 ? _e : {};
        this._x = opts === null || opts === void 0 ? void 0 : opts.xAxisOffset;
        this._y = opts === null || opts === void 0 ? void 0 : opts.yAxisOffset;
        this._tapToDismiss = (_f = opts === null || opts === void 0 ? void 0 : opts.tapToDismiss) !== null && _f !== void 0 ? _f : false;
        this._anchorView = opts === null || opts === void 0 ? void 0 : opts.anchorView;
        // create the android Toast
        this._toast = android.widget.Toast.makeText(Utils.android.getApplicationContext(), this._text, android.widget.Toast.LENGTH_SHORT);
        // const pDrawable = new android.graphics.drawable.PaintDrawable();
        // pDrawable.setCornerRadius(8);
        // const v = this._toast.getView();
        // console.dir(v.findViewById(param0));
        // v.setBackgroundDrawable(pDrawable);
        // console.log('set the toast background to paint drawable');
        const ref = new WeakRef(this);
        (_g = this._toast.getView()) === null || _g === void 0 ? void 0 : _g.setOnTouchListener(new android.view.View.OnTouchListener({
            onTouch(param0, param1) {
                var _a, _b, _c;
                if ((_a = ref.get()) === null || _a === void 0 ? void 0 : _a._tapToDismiss) {
                    (_c = (_b = ref.get()) === null || _b === void 0 ? void 0 : _b._toast) === null || _c === void 0 ? void 0 : _c.cancel();
                }
                return false;
            },
        }));
        try {
            if (android.widget.Toast.Callback) {
                const callback = android.widget.Toast.Callback.extend({
                    owner: null,
                    onToastShown: function () {
                        var _a, _b, _c, _d;
                        this.super.onToastShown();
                        (_d = (_c = (_b = (_a = this.owner) === null || _a === void 0 ? void 0 : _a.get()) === null || _b === void 0 ? void 0 : _b._androidOpts) === null || _c === void 0 ? void 0 : _c.onShown) === null || _d === void 0 ? void 0 : _d.call(_c);
                    },
                    onToastHidden: function () {
                        var _a, _b, _c, _d;
                        this.super.onToastHidden();
                        (_d = (_c = (_b = (_a = this.owner) === null || _a === void 0 ? void 0 : _a.get()) === null || _b === void 0 ? void 0 : _b._androidOpts) === null || _c === void 0 ? void 0 : _c.onHidden) === null || _d === void 0 ? void 0 : _d.call(_c);
                    },
                });
                const cb = new callback();
                cb.owner = ref;
                this._toast.addCallback(cb);
            }
        }
        catch (e) { }
        // set the values
        this.setToastDuration(this._duration).setToastPosition(this._position).setTextColor(this._textColor).setBackgroundColor(this._backgroundColor);
        return this;
    }
    get position() {
        return this._position;
    }
    set position(value) {
        if (value) {
            this._position = value;
            this.setToastPosition(value);
        }
    }
    get duration() {
        return this._duration;
    }
    set duration(value) {
        if (value) {
            this._duration = value;
            this.setToastDuration(value);
        }
    }
    get textColor() {
        return this._textColor;
    }
    set textColor(value) {
        if (value) {
            this._textColor = value;
            this.setTextColor(value);
        }
    }
    get backgroundColor() {
        return this._backgroundColor;
    }
    set backgroundColor(value) {
        if (value) {
            this._backgroundColor = value;
            this.setBackgroundColor(value);
        }
    }
    set xAxisOffset(value) {
        this._x = value;
        this._updateToastPosition();
    }
    set yAxisOffset(value) {
        this._y = value;
        this._updateToastPosition();
    }
    get xAxisOffset() {
        return this._x;
    }
    get yAxisOffset() {
        return this._y;
    }
    static isLength(value) {
        return value && !Utils.isNullOrUndefined(value.unit) && !Utils.isNullOrUndefined(value.value);
    }
    get yPixels() {
        let y;
        if (typeof this._y === 'number') {
            y = this._y * Screen.mainScreen.scale;
        }
        if (Toasty.isLength(this._y)) {
            if (this._y.unit === 'px') {
                y = this._y.value;
            }
            if (this._y.unit === 'dip') {
                y = this._y.value * Screen.mainScreen.scale;
            }
        }
        if (typeof this._y === 'string') {
            if (this._y.includes('px')) {
                y = parseInt(this._y.replace('px', ''), 0);
            }
            if (this._y.includes('dip')) {
                y = parseInt(this._y.replace('dip', ''), 0) * Screen.mainScreen.scale;
            }
        }
        return y;
    }
    get xPixels() {
        let x;
        if (typeof this._x === 'number') {
            x = this._x * Screen.mainScreen.scale;
        }
        if (Toasty.isLength(this._x)) {
            if (this._x.unit === 'px') {
                x = this._x.value;
            }
            if (this._x.unit === 'dip') {
                x = this._x.value * Screen.mainScreen.scale;
            }
        }
        if (typeof this._x === 'string') {
            if (this._x.includes('px')) {
                x = parseInt(this._x.replace('px', ''), 0);
            }
            if (this._x.includes('dip')) {
                x = parseInt(this._x.replace('dip', ''), 0) * Screen.mainScreen.scale;
            }
        }
        return x;
    }
    show() {
        var _a;
        if (!this._text) {
            throw new Error('Toasty Text is not set.');
        }
        else {
            (_a = this._toast) === null || _a === void 0 ? void 0 : _a.show();
        }
    }
    _measureToast() {
        var _a, _b, _c;
        const window = (_a = (Application.android.foregroundActivity || Application.android.startActivity)) === null || _a === void 0 ? void 0 : _a.getWindow();
        const metrics = new android.util.DisplayMetrics();
        window === null || window === void 0 ? void 0 : window.getWindowManager().getDefaultDisplay().getMetrics(metrics);
        const MeasureSpec = android.view.View.MeasureSpec;
        const widthMeasureSpec = MeasureSpec.makeMeasureSpec(metrics.widthPixels, MeasureSpec.UNSPECIFIED);
        const heightMeasureSpec = MeasureSpec.makeMeasureSpec(metrics.heightPixels, MeasureSpec.UNSPECIFIED);
        (_c = (_b = this._toast) === null || _b === void 0 ? void 0 : _b.getView()) === null || _c === void 0 ? void 0 : _c.measure(widthMeasureSpec, heightMeasureSpec);
    }
    get width() {
        var _a, _b, _c;
        this._measureToast();
        return (_c = ((_b = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView()) === null || _b === void 0 ? void 0 : _b.getMeasuredWidth()) / Screen.mainScreen.scale) !== null && _c !== void 0 ? _c : 0;
    }
    get height() {
        var _a, _b, _c;
        this._measureToast();
        return (_c = ((_b = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView()) === null || _b === void 0 ? void 0 : _b.getMeasuredHeight()) / Screen.mainScreen.scale) !== null && _c !== void 0 ? _c : 0;
    }
    cancel() {
        var _a;
        (_a = this._toast) === null || _a === void 0 ? void 0 : _a.cancel();
    }
    setTextColor(value) {
        var _a;
        if (value) {
            this._textColor = value;
            const view = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView();
            const text = view === null || view === void 0 ? void 0 : view.findViewById(android.R.id.message);
            // set the text color
            if (typeof value === 'string') {
                const nativeColor = new Color(value).android;
                text === null || text === void 0 ? void 0 : text.setTextColor(nativeColor);
            }
            else {
                text === null || text === void 0 ? void 0 : text.setTextColor(value.android);
            }
        }
        return this;
    }
    setBackgroundColor(value) {
        var _a;
        if (value) {
            this._backgroundColor = value;
            const view = (_a = this._toast) === null || _a === void 0 ? void 0 : _a.getView();
            if (typeof value === 'string') {
                const nativeColor = new Color(value).android;
                // Gets the actual oval background of the Toast then sets the colour filter
                view === null || view === void 0 ? void 0 : view.getBackground().setColorFilter(nativeColor, android.graphics.PorterDuff.Mode.SRC_IN);
            }
            else {
                view === null || view === void 0 ? void 0 : view.getBackground().setColorFilter(value.android, android.graphics.PorterDuff.Mode.SRC_IN);
            }
        }
        return this;
    }
    setToastDuration(value) {
        var _a, _b, _c;
        switch (value) {
            case ToastDuration.SHORT:
                (_a = this._toast) === null || _a === void 0 ? void 0 : _a.setDuration(android.widget.Toast.LENGTH_SHORT);
                break;
            case ToastDuration.LONG:
                (_b = this._toast) === null || _b === void 0 ? void 0 : _b.setDuration(android.widget.Toast.LENGTH_LONG);
                break;
            default:
                (_c = this._toast) === null || _c === void 0 ? void 0 : _c.setDuration(android.widget.Toast.LENGTH_SHORT);
                break;
        }
        return this;
    }
    setToastPosition(value) {
        this._position = value;
        this._updateToastPosition();
        return this;
    }
    _updateToastPosition() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
        const yOffset = (_a = this.yPixels) !== null && _a !== void 0 ? _a : 0;
        const xOffset = (_b = this.xPixels) !== null && _b !== void 0 ? _b : 0;
        switch (this._position) {
            case ToastPosition.TOP:
                (_c = this._toast) === null || _c === void 0 ? void 0 : _c.setGravity(android.view.Gravity.TOP, xOffset, yOffset);
                break;
            case ToastPosition.TOP_LEFT:
                (_d = this._toast) === null || _d === void 0 ? void 0 : _d.setGravity(android.view.Gravity.TOP | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case ToastPosition.TOP_RIGHT:
                (_e = this._toast) === null || _e === void 0 ? void 0 : _e.setGravity(android.view.Gravity.TOP | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            case ToastPosition.CENTER:
                (_f = this._toast) === null || _f === void 0 ? void 0 : _f.setGravity(android.view.Gravity.CENTER, xOffset, yOffset);
                break;
            case ToastPosition.CENTER_LEFT:
                (_g = this._toast) === null || _g === void 0 ? void 0 : _g.setGravity(android.view.Gravity.CENTER | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case ToastPosition.CENTER_RIGHT:
                (_h = this._toast) === null || _h === void 0 ? void 0 : _h.setGravity(android.view.Gravity.CENTER | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            case ToastPosition.BOTTOM:
                (_j = this._toast) === null || _j === void 0 ? void 0 : _j.setGravity(android.view.Gravity.BOTTOM, xOffset, yOffset);
                break;
            case ToastPosition.BOTTOM_LEFT:
                (_k = this._toast) === null || _k === void 0 ? void 0 : _k.setGravity(android.view.Gravity.BOTTOM | android.view.Gravity.LEFT, xOffset, yOffset);
                break;
            case ToastPosition.BOTTOM_RIGHT:
                (_l = this._toast) === null || _l === void 0 ? void 0 : _l.setGravity(android.view.Gravity.BOTTOM | android.view.Gravity.RIGHT, xOffset, yOffset);
                break;
            default:
                break;
        }
    }
}
