// @triniwiz/nativescript-socketio/angular/bundles/nativescript-nativescript-socketio-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@triniwiz/nativescript-socketio')) :
    typeof define === 'function' && define.amd ? define('nativescript-nativescript-socketio-angular', ['exports', '@angular/core', '@triniwiz/nativescript-socketio'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global['nativescript-nativescript-socketio-angular'] = {}, global.ng.core, global.nativescriptSocketio));
}(this, (function (exports, core, nativescriptSocketio) { 'use strict';

    var SOCKETIO_URL = new core.InjectionToken('SOCKETIO_URL');
    var SOCKETIO_OPTIONS = new core.InjectionToken('SOCKETIO_OPTIONS');
    function socketIOFactory(url, options) {
        return new nativescriptSocketio.SocketIO(url, options);
    }
    var SocketIOModule = /** @class */ (function () {
        function SocketIOModule() {
        }
        SocketIOModule.forRoot = function (url, options) {
            if (options === void 0) { options = {}; }
            return {
                ngModule: SocketIOModule,
                providers: [
                    {
                        provide: nativescriptSocketio.SocketIO,
                        useFactory: socketIOFactory,
                        deps: [SOCKETIO_URL, SOCKETIO_OPTIONS]
                    },
                    { provide: SOCKETIO_URL, useValue: url },
                    { provide: SOCKETIO_OPTIONS, useValue: options },
                ]
            };
        };
        return SocketIOModule;
    }());
    SocketIOModule.decorators = [
        { type: core.NgModule }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.SOCKETIO_OPTIONS = SOCKETIO_OPTIONS;
    exports.SOCKETIO_URL = SOCKETIO_URL;
    exports.SocketIOModule = SocketIOModule;
    exports.socketIOFactory = socketIOFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
