// @nativescript/core/ui/layouts/root-layout/index.ios.js
import { Color } from '../../../color';
import { RootLayoutBase, defaultShadeCoverOptions } from './root-layout-common';
export * from './root-layout-common';
export class RootLayout extends RootLayoutBase {
    constructor() {
        super();
    }
    _bringToFront(view) {
        this.nativeViewProtected.bringSubviewToFront(view.nativeViewProtected);
    }
    _initShadeCover(view, shadeOptions) {
        var _a;
        const initialState = Object.assign(Object.assign({}, defaultShadeCoverOptions.animation.enterFrom), (_a = shadeOptions === null || shadeOptions === void 0 ? void 0 : shadeOptions.animation) === null || _a === void 0 ? void 0 : _a.enterFrom);
        this._applyAnimationProperties(view, initialState);
    }
    _updateShadeCover(view, shadeOptions) {
        return new Promise((resolve) => {
            var _a, _b;
            const options = Object.assign(Object.assign({}, defaultShadeCoverOptions), shadeOptions);
            if (view && view.nativeViewProtected) {
                const duration = this._convertDurationToSeconds(((_b = (_a = options.animation) === null || _a === void 0 ? void 0 : _a.enterFrom) === null || _b === void 0 ? void 0 : _b.duration) || defaultShadeCoverOptions.animation.enterFrom.duration);
                UIView.animateWithDurationAnimationsCompletion(duration, () => {
                    view.nativeViewProtected.backgroundColor = new Color(options.color).ios;
                    this._applyAnimationProperties(view, {
                        translateX: 0,
                        translateY: 0,
                        scaleX: 1,
                        scaleY: 1,
                        rotate: 0,
                        opacity: shadeOptions.opacity,
                    });
                }, (completed) => {
                    resolve();
                });
            }
        });
    }
    _closeShadeCover(view, shadeOptions) {
        return new Promise((resolve) => {
            var _a;
            const exitState = Object.assign(Object.assign({}, defaultShadeCoverOptions.animation.exitTo), (_a = shadeOptions === null || shadeOptions === void 0 ? void 0 : shadeOptions.animation) === null || _a === void 0 ? void 0 : _a.exitTo);
            if (view && view.nativeViewProtected) {
                UIView.animateWithDurationAnimationsCompletion(this._convertDurationToSeconds(exitState.duration), () => {
                    this._applyAnimationProperties(view, exitState);
                }, (completed) => {
                    resolve();
                });
            }
        });
    }
    _applyAnimationProperties(view, shadeCoverAnimation) {
        const translate = CGAffineTransformMakeTranslation(shadeCoverAnimation.translateX, shadeCoverAnimation.translateY);
        // ios doesn't like scale being 0, default it to a small number greater than 0
        const scale = CGAffineTransformMakeScale(shadeCoverAnimation.scaleX || 0.1, shadeCoverAnimation.scaleY || 0.1);
        const rotate = CGAffineTransformMakeRotation((shadeCoverAnimation.rotate * Math.PI) / 180); // convert degress to radians
        const translateAndScale = CGAffineTransformConcat(translate, scale);
        view.nativeViewProtected.transform = CGAffineTransformConcat(rotate, translateAndScale);
        view.nativeViewProtected.alpha = shadeCoverAnimation.opacity;
    }
    _convertDurationToSeconds(duration) {
        return duration / 1000;
    }
}
